src/phantom_token.c

```diff
@@ -0,0 +1,525 @@
+/*
+ *  Copyright 2017 Curity AB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+#include <ngx_string.h>
+#include <stdbool.h>
+#include <assert.h>
+#include "phantom_token.h"
+#include "phantom_token_utils.h"
+
+typedef struct
+{
+    ngx_uint_t done;
+    ngx_uint_t status;
+    ngx_str_t jwt;
+} phantom_token_module_context_t;
+
+static ngx_int_t post_configuration(ngx_conf_t *config);
+
+static void *create_location_configuration(ngx_conf_t *config);
+
+static char *merge_location_configuration(ngx_conf_t *main_config, void *parent, void *child);
+
+static ngx_int_t handler(ngx_http_request_t *request);
+
+static ngx_int_t introspection_response_handler(ngx_http_request_t *request, void *data, ngx_int_t introspection_subrequest_status_code);
+
+static ngx_command_t phantom_token_module_directives[] =
+{
+    {
+        ngx_string("phantom_token"),
+        NGX_HTTP_MAIN_CONF | NGX_HTTP_SRV_CONF | NGX_HTTP_LOC_CONF | NGX_CONF_FLAG,
+        ngx_conf_set_flag_slot,
+        NGX_HTTP_LOC_CONF_OFFSET,
+        offsetof(phantom_token_configuration_t, enable),
+        NULL
+    },
```

Comment:
We no longer configure a client credential or need to remember and then restore headers so boiler plate code is reduced.